main/ejml-dsparse/src/org/ejml/sparse/csc/RandomMatrices_DSCC.java

```diff
@@ -277,6 +278,62 @@ public static DMatrixSparseCSC symmetricPosDef( int width, double probabilityZer
         return out;
     }
 
+    /**
+     * Creates a random matrix where each column has exactly `nzEntriesPerColumn` non-zero entries.
+     * Compared to `rectangle` this method can generate larger sparse matrices.
+     *
+     * @param numRows Number of rows
+     * @param numCols Number of columns
+     * @param nzEntriesPerColumn Amount of nz-entries per column
+     * @param min Minimum element value, inclusive
+     * @param max Maximum element value, inclusive
+     * @param sorted Whether the nz-entries should be sorted
+     * @param rand Random number generator
+     * @return Randomly generated matrix
+     */
+    public static DMatrixSparseCSC generateUniform(int numCols, int numRows, int nzEntriesPerColumn,
+                                                   double min, double max, boolean sorted, Random rand) {
+        if (nzEntriesPerColumn > numRows) {
+            throw new IllegalArgumentException("numRows must be greater than nzEntriesPerColumn");
+        }
+
+        int nz_total = Math.toIntExact(nzEntriesPerColumn * numCols);
+
+        DMatrixSparseCSC matrix = new DMatrixSparseCSC(numCols, numRows, nz_total);
+        matrix.indicesSorted = sorted;
+
+
+        int[] nz_hist = new int[numCols];
+        Arrays.fill(nz_hist, nzEntriesPerColumn);
+        matrix.histogramToStructure(nz_hist);
+
+        BitSet existingRows = new BitSet(numRows);
+
+        for (int i = 0; i < nz_total; i++) {
+            if (i % nzEntriesPerColumn == 0) {
+                existingRows.clear();
+            }
+
+            int row = rand.nextInt(numRows);
+            // avoid duplicate entries
+            while(existingRows.get(row)) {
```

Comment:
This approach to ensuring there are no duplicates can be extremely slow. Imaging if there is only one value left out of a possible million. That's why the shuffle approach was used earlier. Maybe the thing to do is if the probability of selecting a valid value < 5%, create a new subset of values which remain and randomly select from that.

Reply:
Thanks for pointing that out.

I think a better way is to decide whether to chose random values or remove random values. So if more than half of the row indices are to be assigned, it will to remove values.

Removing could be done via a bitset

Reply:
yes that's a much better idea.

Reply:
Will the remove function have a complexity of O(N)?

Reply:
I dont think so.
The row selection is based on a random number generator.
At the worst case, the remove variant has to drop half of the row indices.
Correspondingly, for the select variant, the worst case is to select half of the row indices.

Now the random number generator basically assures, that it wont be `O(N²)` but more close to `O(N)`.
Its a bit vague, but I think its a good apporximation. Maybe we a benchmark could provide more insight (potentially also compared against the `rectangle` version).